ownCloud File Provider/FileProviderExtension.m

```diff
@@ -953,73 +956,123 @@ - (NSProgress *)fetchThumbnailsForItemIdentifiers:(NSArray<NSFileProviderItemIde
 #pragma mark - Core
 - (OCBookmark *)bookmark
 {
+	OCBookmark *bookmark = nil;
+
 	@synchronized(self)
 	{
-		if (_bookmark == nil)
+		bookmark = _bookmark;
+	}
+
+	if (bookmark == nil)
+	{
+		NSFileProviderDomainIdentifier domainIdentifier;
+
+		if ((domainIdentifier = self.domain.identifier) != nil)
 		{
-			NSFileProviderDomainIdentifier domainIdentifier;
+			NSUUID *bookmarkUUID = [[NSUUID alloc] initWithUUIDString:domainIdentifier];
 
-			if ((domainIdentifier = self.domain.identifier) != nil)
-			{
-				NSUUID *bookmarkUUID = [[NSUUID alloc] initWithUUIDString:domainIdentifier];
+			bookmark = [[OCBookmarkManager sharedBookmarkManager] bookmarkForUUID:bookmarkUUID];
 
-				_bookmark = [[OCBookmarkManager sharedBookmarkManager] bookmarkForUUID:bookmarkUUID];
+			if (bookmark == nil)
+			{
+				OCLogError(@"Error retrieving bookmark for domain %@ (UUID %@) - reloading", OCLogPrivate(self.domain.displayName), OCLogPrivate(self.domain.identifier));
 
-				if (_bookmark == nil)
-				{
-					OCLogError(@"Error retrieving bookmark for domain %@ (UUID %@) - reloading", OCLogPrivate(self.domain.displayName), OCLogPrivate(self.domain.identifier));
+				[[OCBookmarkManager sharedBookmarkManager] loadBookmarks];
 
-					[[OCBookmarkManager sharedBookmarkManager] loadBookmarks];
+				bookmark = [[OCBookmarkManager sharedBookmarkManager] bookmarkForUUID:bookmarkUUID];
 
-					_bookmark = [[OCBookmarkManager sharedBookmarkManager] bookmarkForUUID:bookmarkUUID];
+				if (bookmark == nil)
+				{
+					OCLogError(@"Error retrieving bookmark for domain %@ (UUID %@) - final", OCLogPrivate(self.domain.displayName), OCLogPrivate(self.domain.identifier));
+				}
```

Comment:
This code is duplicated with the lines above. Is this needed or may it return a bookmark when calling it twice?

Reply:
The way that method works is by asking the `OCBookmarkManager` for a bookmark with the UUID. If it can't be found, it assumes its list must be out of data and triggers a reload of the bookmarks. Then, if it still can't be found, it logs another error.

Now, looking at it, I think the first log message is better put into the debug level, however, as it really isn't a hard error at this point - only a debug info. I'll change that accordingly.

Reply:
Ok, just committed the change.